File: golang/binary-tree-general/count_complete_tree_nodes.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func countNodes(root *TreeNode) int {
	if root == nil {
		return 0
	}

	left_depth := getDepth(root.Left)
	right_depth := getDepth(root.Right)

	if left_depth == right_depth {
		return 1<<left_depth + countNodes(root.Right)
	} else {
		return 1<<right_depth + countNodes(root.Left)
	}
}

func getDepth(node *TreeNode) int {
	depth := 0

	for node != nil {
		depth += 1

		node = node.Left
	}

	return depth
}

File: python/binary-tree-general/count_complete_tree_nodes.py
from typing import Optional


class TreeNode:
    def __init__(self, val: int = 0, left: Optional['TreeNode'], right: Optional['TreeNode']):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def countNodes(self, root: TreeNode) -> int:
        if not root:
            return 0

        left_depth = self.getDepth(root.left)
        right_depth = self.getDepth(root.right)

        if left_depth == right_depth:
            return (1 << left_depth)+ self.countNodes(root.right)
        else:
            return (1 << right_depth) + self.countNodes(root.left)



    def getDepth(self, node: Optional[TreeNode])  -> int:
        depth = 0

        while node:
            depth += 1

            node = node.left

        return depth
